### PR TITLE
copy: introduce copy-buffer-kilobytes build option

### PR DESCRIPTION
**Pull request by AndreiSva**
### Description
This pull request adds a new entry to `meson_options.txt`, allowing the build-time configuration of the size of the buffer used in `src/shared/copy.c`, in kilobytes. The default value of the option remains 64 KiB and the option is configured as to not allow values smaller than 1 KiB.

I decided that it would be sensible to have the configuration be set in kilobytes, as it makes changing the default value or otherwise configuring it more streamlined than specifying the value directly in bytes.

### Rational
- This patch facilitates more streamlined deployment of high-performance systemd Linux systems on computers with large amounts of RAM.  COPY_BUFFER_SIZE as defined in `src/shared/copy.c` is indirectly used in many parts of systemd, thus being able to increase it to a larger value can drastically increase performance across many areas of the system as shown in 9b2d9b07a17158b3a2830e1b77ae77534eed3787. 

- Conversely this patch also makes it easier to deploy custom systemd builds on embedded platforms or computers with very little RAM. Although the overall memory footprint of the buffer might seem negligible, many concurrent calls to utilities that depend on `copy.c` (such as systemd-nspawn) could conceivably add up to several megabytes, which is significant in embedded applications. 

In conclusion, I believe that this change is sensible and would be beneficial in maintaining systemd's versatility and its wide range of deployment scenarios.

### CODE REVIEW

**Comment by bluca** [comment]
I don't think there should be a build option for this, it's way too specific. If anything, some heuristic based on total memory available should be acceptable - ie, if memory is below X, then set to Y. Can you provide some sensible values for the heuristic to use?

**Comment by AndreiSva** [comment]
> I don't think there should be a build option for this, it's way too specific. If anything, some heuristic based on total memory available should be acceptable - ie, if memory is below X, then set to Y. Can you provide some sensible values for the heuristic to use?

That is a very good point. I think using that sort of model would also have the benefit of potentially being applicable across different components across systemd in the future.

My opinion on the potential heuristic values is that it would be a good idea to have separate cases for systems with 256, 512, and 1024 MiB of RAM. For example, in regard to the copy buffer, systems with 256, 512, and 1024 MiB could allocate 8 KiB, 16 KiB, and 32 KiB, respectively. As for the upper end, I am undecided, as our specific use case is focused on getting the most out of our 512 MiB ARM board. Maybe doubling the size of the buffer again at 2048 or 4096 MiB would be reasonable, and allocating a maximum of around 256 KiB on machines with a lot of RAM. Linux has a pretty big maximum stack size, so I think there is a fair bit of flexibility.

I could also see people disagreeing on how they want this handled, so maybe it would still be a good idea to have the values be configurable in some way.

**Comment by poettering** [comment]
uh, sorry, but this should not be a configurable knob, and i doubt a much larger buffer will do you much benefit. The larger you make buffers the smaller the effect will become. Moreover you will eventually increase memory pressure, evicting more useful data from memory needlessly. This buffer is also allocated on the stack there overly large allocations should not be placed.

If you can provide profiling data that a larger buffer really has a non-trivial positive effect, then I am happy to increase the buffer – for everyone. But without profiling data this is just poking in the fog.

**Comment by bluca** [comment]
I think the intent here is using a smaller one on small systems, so a runtime check for total system memory that adjusts it down when on really small systems would make sense

**Comment by AndreiSva** [comment]
> I think the intent here is using a smaller one on small systems, so a runtime check for total system memory that adjusts it down when on really small systems would make sense

Yes, my focus is very much on the lower-end of things. I think this feature would be really valuable. The effect on more powerful machines was mostly tentative, though I thought it was a thing worth bringing up / discussing.

**Comment by AndreiSva** [comment]
@bluca I closed this for now because the original idea is not ideal, but I would still want to see the heuristic feature implemented, so I might open a new PR. What do you think the way forward with this is?

**Comment by bluca** [comment]
Feel free to open a new PR with the new feature
